src/data_source/fs.rs

```diff
@@ -0,0 +1,673 @@
+/// # use hotshot_query_service::testing::mocks::{
+/// #   MockNodeImpl as AppNodeImpl, MockTypes as AppTypes,
+/// # };
+/// # use std::path::Path;
+/// # use tide_disco::App;
+/// # type AppQueryData = ();
+/// struct AppState {
+///     // Top-level storage coordinator
+///     store: AtomicStore,
+///     hotshot_qs: FileSystemDataSource<AppTypes, AppNodeImpl, AppQueryData>,
+///     // additional state for other modules
+/// }
+///
+/// fn init_server(
+///     storage_path: &Path,
+///     mut hotshot: SystemContextHandle<AppTypes, AppNodeImpl>,
+/// ) -> Result<App<Arc<RwLock<AppState>>, Error>, Error> {
+///     let mut loader = AtomicStoreLoader::create(storage_path, "my_app") // or `open`
+///         .map_err(Error::internal)?;
+///     let hotshot_qs = FileSystemDataSource::create_with_store(
+///         &mut loader,
+///         AppQueryData::default(),
+///     )
+///     .map_err(Error::internal)?;
+///     // Initialize storage for other modules using the same loader.
+///
+///     let store = AtomicStore::open(loader).map_err(Error::internal)?;
+///     let state = Arc::new(RwLock::new(AppState {
+///         store,
+///         hotshot_qs,
+///         // additional state for other modules
+///     }));
+///     let mut app = App::with_state(state.clone());
+///     // Register API modules.
+///
+///     spawn(async move {
+///         let mut events = hotshot.get_event_stream(Default::default()).await.0;
+///         while let Some(event) = events.next().await {
+///             let mut state = state.write().await;
+///             state.hotshot_qs.update(&event).await.unwrap();
+///             // Update other modules' states based on `event`.
+///
+///             state.hotshot_qs.commit_version().await.unwrap();
+///             // Commit or skip versions for other modules' storage.
+///             state.store.commit_version().unwrap();
+///         }
+///     });
+///
+///     Ok(app)
+/// }
+/// ```
+#[derive(custom_debug::Debug)]
+pub struct FileSystemDataSource<Types: NodeType, I: NodeImplementation<Types>, UserData>
+where
+    Block<Types>: QueryableBlock,
+{
+    index_by_leaf_hash: HashMap<LeafHash<Types, I>, u64>,
+    index_by_block_hash: HashMap<BlockHash<Types>, u64>,
+    index_by_txn_hash: HashMap<TransactionHash<Types>, (u64, TransactionIndex<Types>)>,
+    index_by_proposer_id: HashMap<EncodedPublicKey, Vec<u64>>,
+    #[debug(skip)]
+    top_storage: Option<AtomicStore>,
+    leaf_storage: LedgerLog<LeafQueryData<Types, I>>,
+    block_storage: LedgerLog<BlockQueryData<Types>>,
+    metrics: PrometheusMetrics,
+    user_data: UserData,
+}
+
+impl<Types: NodeType, I: NodeImplementation<Types>, UserData>
+    FileSystemDataSource<Types, I, UserData>
+where
+    Block<Types>: QueryableBlock,
+{
+    /// Create a new [FileSystemDataSource] with storage at `path`.
+    ///
+    /// If there is already data at `path`, it will be archived.
```

Comment:
This isn't part of this PR so we can maybe just create an issue for it if we want to fix it. 

I think it's better to fail if one calls `create` on a non-empty directory because it's likely a user error. It should be easy for the user of the code to make sure it's using an empty or non-existent directory. It's somewhat harder to figure out what files to move where if they were moved by mistake.